### ISSUE TITLE
TiledMap Rendering Issue

### ISSUE THREAD

**Issue by stefanrbk** [craftworkgames/MonoGame.Extended#476]
I have discovered some sort of rendering issue with the TiledMapRenderer or its interaction with the BoxingViewportAdapter. It appears that as the tiles in the layers are being drawn, the left-most column and top-most row of pixels per tile are being cut off.
![1](https://user-images.githubusercontent.com/23112788/36454912-874c2102-1663-11e8-98ef-5d4945f546a7.PNG)
(I added the red lines to show the pixel boundaries)
It is difficult for me to see on the full map, but it does look like this occurs with every tile! I have been spinning in circles digging through the TiledMapRenderer, BoxingViewportAdapter and it's parent ViewportAdapter classes and I have no idea where this rendering error is coming from. I am using a TiledMap and a BoxingViewportAdapter of size 240x160 (GBA screen size). Anyone have any ideas what is going on?

**Comment by craftworkgames** [comment]
What you do mean by GBA screen size?

This looks like a scaling issue. Chances are there's some sort of scaling being applied by the viewport adapter or the camera as a view matrix to the map.

**Comment by stefanrbk** [comment]
GameBoy Advance. I'm building a Pokemon clone to test out the engine.

Here is the scaling code I am using. 

Startup
`ViewportAdapter = new BoxingViewportAdapter(Window, GraphicsDevice, __width, __height);`
`_camera = new Camera2D(ViewportAdapter);`
(__width and __height are the screen size constants of 240x160)

Draw
`var viewMatrix = _camera.GetViewMatrix();`
`var projectionMatrix = Matrix.CreateOrthographicOffCenter(0, .ViewportAdapter.ViewportWidth, .ViewportAdapter.ViewportHeight, 0, 0f, -1f);`
`_mapRenderer.Draw(_map.Layers[i], ref viewMatrix, ref projectionMatrix);`
(and yes, I am manually drawing each layer separately)

Anything you think could be on my end?

**Comment by craftworkgames** [comment]
What about the PreferredBackBufferWidth / Height of the GraphicsDevice?

**Comment by stefanrbk** [comment]
I manually stretched the window across my screen (as maximize is currently broken when it comes to scaling until MonoGame fixes it). The prefered width/height is 1408x2549. My virtual width/height is 240x160 (ratio of 3:2) and the BoxingViewportAdapter correctly "boxes" its width/height to 1408x2112 (ratio of 3:2 as well).

**Comment by craftworkgames** [comment]
Try using the `DefaultViewportAdapter` and see what happens?  This adapter shouldn't apply any scaling at all so you can at least rule out (or in) the viewport adapter.

I actually think we need a another option (either on the boxing adapter or a new adapter) that allows pixel perfect boxing on both sides. I think [Nez](https://www.youtube.com/watch?v=pl_5sU5izdw) supports something like this but we never got around to implementing it.

**Comment by stefanrbk** [comment]
With the `DefaultViewportAdapter`, everything looks right when stationary since the pixels are rendered in a 1-to-1 way, but that all changes when the camera starts moving.

It looks like a ripple effect. When the camera is moving left/right, it almost looks like all the tiles are bouncing left/right. It's kinda hard to describe and I can't figure out how to screenshot a moving image... I'll move right, and it looks like the tile columns grow 1 pixel wider, then snap back to place.

What I can say is that it looks like the `SpriteBatch` is completely unaffected though. i haven't seen any artifacts with the character I have drawn on the screen. I probably wouldn't have noticed the rendering issue if it was also affected. when rendering 240x160 at 2k resolution, you would expect your pixels to line up!
![image](https://user-images.githubusercontent.com/23112788/36464295-ac3d7638-1694-11e8-816d-3598b112442d.png)

**Comment by craftworkgames** [comment]
Ah okay. Well at least we seem to be getting somewhere.

Take a look at [this line](https://github.com/craftworkgames/MonoGame.Extended/blob/develop/Source/Demos/Demo.Features/Demos/TiledMapsDemo.cs#L167) of the Tiled Map demo. 

Are you setting `GraphicsDevice.SamplerStates[0]` to `SamplerState.PointClamp`?

We had lots of these kinds of issues in the early days of the Tiled renderer but I'm pretty sure they are all gone now. The API is still a little clunky, which is something I'd like to improve but I think we can get it working.

**Comment by stefanrbk** [comment]
Yeah, I did set it. Here is the full drawing method I have cooked up...

And since the code tool on here won't let me do what I want it to do...
![image](https://user-images.githubusercontent.com/23112788/36465174-9a78d0d2-1699-11e8-9146-5a08293adc52.png)

The cut off line is
`var projectionMatrix = Matrix.CreateOrthographicOffCenter(0, _game.ViewportAdapter.ViewportWidth, _game.ViewportAdapter.ViewportHeight, 0, 0f, -1f);`

**Comment by stefanrbk** [comment]
```C#
        public override void Draw(GameTime gameTime)
        {
            _game.GraphicsDevice.Clear(Color.Black);

            _game.GraphicsDevice.BlendState = BlendState.AlphaBlend;
            _game.GraphicsDevice.SamplerStates[0] = SamplerState.PointClamp;
            _game.GraphicsDevice.RasterizerState = RasterizerState.CullNone;

            var viewMatrix = _camera.GetViewMatrix();
            var projectionMatrix = Matrix.CreateOrthographicOffCenter(0, _game.ViewportAdapter.ViewportWidth, _game.ViewportAdapter.ViewportHeight, 0, 0f, -1f);

            int i = 0;
            while (i < _map.Layers.Count)
            {
                if (String.Compare(_map.Layers[i].Name, 0, "~", 0, 1) == 0) break;

                _mapRenderer.Draw(_map.Layers[i++], ref viewMatrix, ref projectionMatrix);
            }

            _spriteBatch.Begin(transformMatrix: viewMatrix, samplerState: SamplerState.PointClamp);
            _spriteBatch.Draw(_heroSprite);
            _spriteBatch.End();

            while (i < _map.Layers.Count)
            {
                if (String.Compare(_map.Layers[i].Name, 0, "~", 0, 1) == 1) break;

                _mapRenderer.Draw(_map.Layers[i++], ref viewMatrix, ref projectionMatrix);
            }

            base.Draw(gameTime);
        }
```
Nope, code tool is working just fine. I must have momentarily caught dumb and forgot how to delineate code... oops

**Comment by craftworkgames** [comment]
Try this:

```csharp
public override void Draw(GameTime gameTime)
{
    _game.GraphicsDevice.Clear(Color.Black);

    var viewMatrix = _camera.GetViewMatrix();
    var projectionMatrix = Matrix.CreateOrthographicOffCenter(0, _game.ViewportAdapter.ViewportWidth, _game.ViewportAdapter.ViewportHeight, 0, 0f, -1f);

    int i = 0;
    while (i < _map.Layers.Count)
    {
        if (String.Compare(_map.Layers[i].Name, 0, "~", 0, 1) == 0) break;

        _mapRenderer.Draw(_map.Layers[i++], ref viewMatrix, ref projectionMatrix);
    }

    _spriteBatch.Begin(transformMatrix: viewMatrix, samplerState: SamplerState.PointClamp);
    _spriteBatch.Draw(_heroSprite);
    _spriteBatch.End();

    _game.GraphicsDevice.BlendState = BlendState.AlphaBlend;
    _game.GraphicsDevice.SamplerStates[0] = SamplerState.PointClamp;
    _game.GraphicsDevice.RasterizerState = RasterizerState.CullNone;

    while (i < _map.Layers.Count)
    {
        if (String.Compare(_map.Layers[i].Name, 0, "~", 0, 1) == 1) break;

        _mapRenderer.Draw(_map.Layers[i++], ref viewMatrix, ref projectionMatrix);
    }

    base.Draw(gameTime);
}
```

My suspicion is that the call to `SpriteBatch` might be overriding those settings.

**Comment by stefanrbk** [comment]
Nope, still there. I tested with both Boxing and Default ViewportAdapters. :tired_face:

**Comment by craftworkgames** [comment]
Bummer.. Well I'm out of ideas for now. I'll have to try an reproduce this issue later.

One last question though? Are you rendering directly on the GBA or are you rendering in a Window using OpenGL or DirectX on Windows?

**Comment by stefanrbk** [comment]
Right, I am rendering on a Windows 10 PC, I'm just trying to make it look like the GameBoy Advance game it is coming from. and i'm unsure which rendering engine i'm using... I believe it to be DirectX?? :sweat_smile:

**Comment by LithiumToast** [comment]
@stefanrbk Did you check that the viewport adapter is scaling by whole integers? You can also try rendering the unscaled scene to a `RenderTarget2D` then drawing that render target with `SpriteBatch` to the dimensions of the back buffer to scale up instead.

**Comment by stefanrbk** [comment]
It is scaling with whole integers. When the window opens, it renders at a scale of 3x and i can even tell then without making the window bigger that it is still there. I have been trying to get the `RenderTarget2D` to work to make it appear scaled as before to see if that would work, but I can't seem to get it working at the moment.

**Comment by stefanrbk** [comment]
I just pulled up the TiledMap demo, I originally changed the render settings to the same as my game's settings, and the render glitch was there. Then I reset the render settings back to the original, and it was still there. The glitch is present in the demo. 
![image](https://user-images.githubusercontent.com/23112788/36515591-c9b7be48-173f-11e8-988e-5b5eee8a30ea.png)
(look at the vertical column in the middle)
To see this in action, launch the Demo.Features demo and select TiledMap. Find the area posted below and hold `R` to zoom in where the grass meets the ground directly under where the background ends and the Blue static background meet.
![image](https://user-images.githubusercontent.com/23112788/36515726-68f30f30-1740-11e8-9992-261a2bb86102.png)

While zooming in and out, there are other artifacts appearing as well. See below.
![image](https://user-images.githubusercontent.com/23112788/36515775-a460d1d8-1740-11e8-9885-e3545ea8c649.png)

**Comment by stefanrbk** [comment]
It looks like this is not an issue with DirectX vs OpenGL. I ran the OpenGL Demo and it also has the gimpy pixels in the same spot, so I don't think there is an issue on that end. I'm trying to wrap my head around how the `TiledMapRenderer` does it's thing. I ran the Graphics debugger and I was able to check some of the primitives, and they have the proper coordinates (2 triangle primitives splitting each tile horizontally from top-right to bottom-left with vertices on multiples of the tile sizes). Only thing I haven't checked yet is how the textures are being mapped to the primitives. Maybe there is a scaling issue there or a `float` rounding error in there which could cause this.

**Comment by craftworkgames** [comment]
@stefanrbk thanks for looking into this. We've had issues like this before and I thought they were all solved. Obviously not.

**Comment by stefanrbk** [comment]
I'm running with `2.0` and it also has the same problem out the gate. I removed the `0.5f` and the skinny pixel went away. I'm not seeing any of the gaps which were present before, but that's probably because i'm using the `PointClamp` as suggested in your post. The tile bleeding still occurs without the `0.5f` when zooming with the camera, so I'm wondering if maybe there is some sort of issue with the camera scaling? According to the graphics debugging and selecting a missing pixel between the tiles shown above, the model with the bleeding was part of the tile to the left of the seam. I will include some diagnostic info.

**View Matrix**
```
4.53859329,0,0,-8318.529
0,4.53777933,0,-1216.29321
0,0,1,0
0,0,0,1
```

**Projection Matrix**
```
0.00107581719,0,0,-1
0,-0.00143472024,0,1
0,0,1,0
0,0,0,1
```

**Buffer Size**
`2527,1393`

**Test Pixel**
`1208,526`

**Triangle**
`37 of 144`

**Vertices**
*73*
Position `(2048,384,0)`
Texture `(0.797546,0.3350384)`
*75*
Position `(2048,512,0)`
Texture `(0.797546,0.4987212)`
*74*
Position `(1920,512,0)`
Texture `(0.601227,0.4987212)`

**Color from texture**
`ARGB(0,0,0,0)`

Not sure if the memdump will help, but those are the values I could pull showing that issue.

**Comment by craftworkgames** [comment]
Fixed with #480
